megamek/src/megamek/client/ui/swing/util/TurnTimer.java

```diff
@@ -0,0 +1,136 @@
+ * details.
+ */
+package megamek.client.ui.swing.util;
+
+
+import megamek.client.Client;
+import megamek.client.ui.swing.AbstractPhaseDisplay;
+import megamek.client.ui.swing.GUIPreferences;
+import megamek.common.IGame;
+import megamek.common.options.Option;
+
+import javax.swing.*;
+import java.awt.*;
+import java.awt.event.ActionEvent;
+import java.awt.event.ActionListener;
+
+/**
+ * This class takes a time limit, which is to be set in Basic Options and counts down to zero
+ * When zero is reached, the ready() method of the given {@link AbstractPhaseDisplay} is called
+ * to end the users current turn.
+ *
+ */
+public class TurnTimer {
+
+    private Timer timer;
+    private JProgressBar progressBar;
+    private ActionListener listener;
+    private JLabel remaining;
+    private JPanel display;
+    private int timeLimit;
+    private AbstractPhaseDisplay phaseDisplay;
+
+
+    public TurnTimer(int limit, AbstractPhaseDisplay pD) {
+
+        phaseDisplay = pD;
+        // make it minutes here.
+        timeLimit = limit*60;
+
+        display = new JPanel();
+        progressBar = new JProgressBar(JProgressBar.HORIZONTAL, 0, timeLimit);
+        progressBar.setValue(timeLimit);
+
+
+        listener = new ActionListener() {
+            int counter = timeLimit;
+            public void actionPerformed(ActionEvent ae) {
+                counter--;
+                int seconds =counter % 60;
+                int minutes = counter/60;
+                remaining.setText(minutes+":"+seconds);
+                progressBar.setValue(counter);
+
+                if (counter<1) {
+                    // get the NagForNoAction setting here
+                    boolean nagSet =   GUIPreferences.getInstance().getNagForNoAction();
+                    // prevent the popup dialog from breaking time limit
+                    GUIPreferences.getInstance().setNagForNoAction(false);
+                    phaseDisplay.ready();
+                    // reset NagForNoAction to the value it was before to preserve the user experience
+                    // for use cases outside the timer
+                    GUIPreferences.getInstance().setNagForNoAction(nagSet);
+                    timer.stop();
+                    display.setVisible(false);
+                    phaseDisplay.getClientgui().getMenuBar().remove(display);
+                   // display.dispose();
+                }
+            }
+        };
+
+    }
+
+    public void startTimer() {
+
+        SwingUtilities.invokeLater( new Runnable() {
+            public void run() {
+                remaining = new JLabel(timeLimit+" sec");
+                timer = new Timer(1000, listener);
+                phaseDisplay.getClientgui().getMenuBar().add(display);
+                timer.start();
+                display.setLayout(new FlowLayout());
+                display.add(remaining);
+                display.add(progressBar);
+                display.setVisible(true);
```

Comment:
There's no need to rebuild the timer panel each time it's displayed. It should be initialized once in the constructor and just set to visible or invisible as needed.

Reply:
I'm sorry it's taking me so long to get back to you. I have so little time these days.

Reply:
done